### PR TITLE
Add sidecar egress hosts conflict warnings

### PR DESCRIPTION
**Pull request by kebe7jun**
For https://github.com/istio/istio/pull/33766#issuecomment-872667419

### CODE REVIEW

**Comment by ramaraochavali** [comment]
LGTM. Will leave it for others to review

**Review by howardjohn — CHANGES_REQUESTED** [review]
Making previously valid configurations invalid is a breaking API change. At most we can warn.

**Comment by kebe7jun** [comment]
> Making previously valid configurations invalid is a breaking API change. At most we can warn.

@howardjohn  Is this an indication that we only need to print the logs?

**Comment by hzxuzhonghu** [comment]
Add a warning, and add a TODO maybe in 1.12 we can make it this way.

FYI: `WrapWarning`

**Comment by kebe7jun** [comment]
> Add a warning, and add a TODO maybe in 1.12 we can make it this way.
> 
> FYI: `WrapWarning`

Thanks for your suggestions, I did it.

**Comment by howardjohn** [comment]
> Add a warning, and add a TODO maybe in 1.12 we can make it this way.

We can never make it this way without it being a breaking API change. To do this it would require a new ApiVersion

File: pkg/config/validation/validation_test.go [hunk 1]
```diff
@@ -5580,10 +5580,53 @@ func TestValidateSidecar(t *testing.T) {
...
+			Egress: []*networking.IstioEgressListener{
+				{
+					Hosts: []string{
+						"test/*",
+						"test/a.com",
+					},
+				},
+			},
+		}, false},
+		{"skip(1.11): sidecar egress duplicated with wildcarded same namespace .", &networking.Sidecar{
```
Comment (howardjohn) [inline pkg/config/validation/validation_test.go:5613]: what does "skip(1.11)" mean?

Comment (howardjohn) [reply pkg/config/validation/validation_test.go:5613]: Oh I see. We should test the warnings as well, using checkValidationMessage like some other tests

**Comment by hzxuzhonghu** [comment]
@howardjohn https://kubernetes.io/docs/reference/using-api/deprecation-policy/

**Comment by howardjohn** [comment]
Yes, I would think this violates "Rule #1: API elements may only be removed
by incrementing the version of the API group"

Removing an element (either key or legal value) and making it invalid are
effectively the same?

**Comment by kebe7jun** [comment]
@howardjohn  I see the same action is allowed in Kubernetes: https://github.com/kubernetes/kubernetes/pull/103245.

**Comment by howardjohn** [comment]
Looks like that may be for an alpha feature flag protected field?

**Comment by kebe7jun** [comment]
> Looks like that may be for an alpha feature flag protected field?

Yes, but this one does not affect the API and resources, it is only print some warnings in pilot, I can not do this also?

**Comment by kebe7jun** [comment]
> Looks like that may be for an alpha feature flag protected field?

Yes, but this one does not affect the API and resources, it is only print some warnings in pilot, can we not do this either?

**Comment by hzxuzhonghu** [comment]
LGTM, it only add some warning in validation, does not break any old configs

**Comment by howardjohn** [comment]
Sorry I either misread it or didn't look at the update - adding a warning is good, just not error

**Review by hzxuzhonghu — APPROVED** [review]

**Pull request merged**